featurebyte/models/mixin.py

```diff
@@ -20,9 +22,14 @@ class QueryGraphMixin(FeatureByteBaseModel):
 
     # special handling for those attributes that are expensive to deserialize
     # internal_* is used to store the raw data from persistence, _* is used as a cache
-    internal_graph: Any = Field(default=None, allow_mutation=False, alias="graph")
+    internal_graph: Any = Field(default=None, frozen=True, alias="graph")
     _graph: Optional[QueryGraph] = PrivateAttr(default=None)
 
+    @field_serializer("internal_graph", when_used="json")
+    def _serialize_graph(self, graph: Any) -> Any:
+        _ = graph
+        return json.loads(self.graph.model_dump_json(by_alias=True))
```

Comment:
To help my understanding, could you clarify what does a field serializer do and when should a model use one in our context?

Reply:
`json` output should not contain type like `ObjectId` and `datetime`. Pydantic V2 deprecates `json_encoders` due to performance issue. We need to declare an explicit `field_serializer` method to serialize a particular field.